File: src/main/java/com/kgaft/VulkanLib/Device/Image/Image.java
package com.kgaft.VulkanLib.Device.Image;

import com.kgaft.VulkanLib.Device.Buffer.Buffer;
import com.kgaft.VulkanLib.Device.LogicalDevice.LogicalDevice;
import com.kgaft.VulkanLib.Device.LogicalDevice.LogicalQueue;
import com.kgaft.VulkanLib.Utils.DestroyableObject;
import com.kgaft.VulkanLib.Utils.LwjglObject;
import com.kgaft.VulkanLib.Utils.VkErrorException;
import org.lwjgl.vulkan.*;

import java.lang.instrument.IllegalClassFormatException;
import java.util.ArrayList;

import static org.lwjgl.vulkan.VK10.*;

public class Image extends DestroyableObject {
    private LogicalDevice device;
    private long base;
    private LwjglObject<VkImageCreateInfo> imageInfo;
    private ArrayList<ImageView> imageViews = new ArrayList<>();
    private long imageMemory;
    private boolean castCreated = false;
    private LwjglObject<VkBufferImageCopy.Buffer> region = new LwjglObject<>(VkBufferImageCopy.class, VkBufferImageCopy.Buffer.class, 1);
    public Image(LogicalDevice device, long base) throws IllegalClassFormatException {
        castCreated = true;
        this.device = device;
        this.base = base;
    }

    public Image(LogicalDevice device, LwjglObject<VkImageCreateInfo> createInfo) throws VkErrorException, IllegalClassFormatException {
        initialize(device, createInfo);
    }

    public Image() throws IllegalClassFormatException {
    }

    public void initialize(LogicalDevice device, long image) {
        this.device = device;
        this.base = image;
    }

    public void initialize(LogicalDevice device, LwjglObject<VkImageCreateInfo> createInfo) throws VkErrorException {
        this.imageInfo = createInfo;
        this.device = device;
        long[] res = new long[1];
        int status = vkCreateImage(device.getDevice(), createInfo.get(), null, res);
        if (status != VK_SUCCESS) {
            throw new VkErrorException("Failed to create image", status);
        }
        this.base = res[0];

        try {
            LwjglObject<VkMemoryRequirements> reqs = new LwjglObject<>(VkMemoryRequirements.class);

            vkGetImageMemoryRequirements(device.getDevice(), base, reqs.get());
            LwjglObject<VkMemoryAllocateInfo> allocInfo = new LwjglObject<>(VkMemoryAllocateInfo.class);
            allocInfo.get().allocationSize(reqs.get().size());
            allocInfo.get().memoryTypeIndex(device.findMemoryType(reqs.get().memoryTypeBits(), VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT));

            status = vkAllocateMemory(device.getDevice(), allocInfo.get(), null, res);
            if (status != VK_SUCCESS) {
                throw new VkErrorException("Failed to allocate image memory", status);
            }
            this.imageMemory = res[0];

        } catch (IllegalClassFormatException e) {
            throw new RuntimeException(e);
        }
    }

    public ImageView createImageView(LwjglObject<VkImageViewCreateInfo> createInfo) throws VkErrorException {
        long[] res = new long[1];
        int status = vkCreateImageView(device.getDevice(), createInfo.get(), null, res);
        if (status != VK_SUCCESS) {
            throw new VkErrorException("Failed to create image view: ", status);
        }
        ImageView result = new ImageView(device, res[0], createInfo, this.imageInfo);
        imageViews.add(result);
        return result;
    }


    public void transitionImageLayout(LogicalDevice device, int oldLayout, int newLayout,
                                      int aspectFlags) {
        VkCommandBuffer commandBuffer = device.getQueueByType(VK_QUEUE_GRAPHICS_BIT).beginCommandBuffer();

        transitionImageLayout(commandBuffer, oldLayout, newLayout, aspectFlags);

        device.getQueueByType(VK_QUEUE_GRAPHICS_BIT).endSingleTimeCommands(commandBuffer);
    }

    public void transitionImageLayout(VkCommandBuffer commandBuffer, int oldLayout, int newLayout,
                                      int aspectFlags) {
        try {
            LwjglObject<VkImageMemoryBarrier.Buffer> barrier = new LwjglObject<>(VkImageMemoryBarrier.class, VkImageMemoryBarrier.Buffer.class, 1);
            barrier.get().oldLayout(oldLayout);
            barrier.get().newLayout(newLayout);
            barrier.get().srcQueueFamilyIndex(VK_QUEUE_FAMILY_IGNORED);
            barrier.get().dstQueueFamilyIndex(VK_QUEUE_FAMILY_IGNORED);
            barrier.get().image(base);
            barrier.get().subresourceRange().aspectMask(aspectFlags);
            barrier.get().subresourceRange().baseMipLevel(0);
            barrier.get().subresourceRange().levelCount(1);
            barrier.get().subresourceRange().baseArrayLayer(0);
            barrier.get().subresourceRange().layerCount(1);

            int sourceStage;
            int destinationStage;

            if (oldLayout == VK_IMAGE_LAYOUT_UNDEFINED && newLayout == VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL) {
                barrier.get().srcAccessMask(0);
                barrier.get().dstAccessMask(VK_ACCESS_TRANSFER_WRITE_BIT);

                sourceStage = (VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT);
                destinationStage = VK_PIPELINE_STAGE_TRANSFER_BIT;
            } else if (oldLayout == VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL &&
                    newLayout == VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL) {
                barrier.get().srcAccessMask(VK_ACCESS_TRANSFER_WRITE_BIT);
                barrier.get().dstAccessMask(VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT);

                sourceStage = VK_PIPELINE_STAGE_TRANSFER_BIT;
                destinationStage = VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
            } else {
                barrier.get().srcAccessMask(0);
                barrier.get().dstAccessMask(VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT);

                sourceStage = VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT;
                destinationStage = VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
            }

            vkCmdPipelineBarrier(
                    commandBuffer,
                    sourceStage, destinationStage,
                    0,
                    null,
                    null, barrier.get());

        } catch (Exception e) {
            throw new RuntimeException(e);
        }

    }
    public void copyFromBuffer(Buffer buffer, int layerCount, LogicalQueue queue) {
        VkCommandBuffer cmd = queue.beginCommandBuffer();
        transitionImageLayout(cmd, imageInfo.get().initialLayout(), VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, VK_IMAGE_ASPECT_COLOR_BIT);

        region.get().bufferOffset(0);
        region.get().bufferRowLength(0);
        region.get().bufferImageHeight(0);
        region.get().imageSubresource().aspectMask(VK_IMAGE_ASPECT_COLOR_BIT);
        region.get().imageSubresource().mipLevel(0);
        region.get().imageSubresource().baseArrayLayer(0);
        region.get().imageSubresource().layerCount(layerCount);
        region.get().imageOffset().x(0);
        region.get().imageOffset().y(0);
        region.get().imageOffset().z(0);
        region.get().imageExtent(imageInfo.get().extent());
        vkCmdCopyBufferToImage(cmd, buffer.getBuffer(), base, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, region.get());
        transitionImageLayout(cmd,VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL,  VK_IMAGE_LAYOUT_GENERAL, VK_IMAGE_ASPECT_COLOR_BIT);
        imageInfo.get().initialLayout(VK_IMAGE_LAYOUT_GENERAL);
        queue.endSingleTimeCommands(cmd);
    }

    public void resize(int width, int height) throws VkErrorException {
        if (!castCreated) {
            destroy();
            destroyed = false;
            imageInfo.get().extent().width(width);
            imageInfo.get().extent().height(height);
            imageInfo.get().extent().depth(1);
            if(imageInfo.get().initialLayout()!=VK_IMAGE_LAYOUT_UNDEFINED){
                int tmpLayout = imageInfo.get().initialLayout();
                imageInfo.get().initialLayout(VK_IMAGE_LAYOUT_UNDEFINED);
                initialize(device, imageInfo);
                transitionImageLayout(device, VK_IMAGE_LAYOUT_UNDEFINED, tmpLayout, VK_IMAGE_ASPECT_COLOR_BIT);
                imageInfo.get().initialLayout(tmpLayout);
            } else {
                initialize(device, imageInfo);
            }
            imageViews.forEach(element->{
                element.createInfo.get().image(base);
                long[] res = new long[1];
                vkCreateImageView(device.getDevice(), element.createInfo.get(), null, res);
                element.imageView = res[0];
                element.parentInfo = imageInfo;
            });

        }

    }

    public long getBase() {
        return base;
    }

    public LwjglObject<VkImageCreateInfo> getImageInfo() {
        return imageInfo;
    }

    public ArrayList<ImageView> getImageViews() {
        return imageViews;
    }

    @Override
    public void destroy() {
        if(!castCreated){
            imageViews.forEach(ImageView::destroy);
            vkDestroyImage(device.getDevice(), base, null);
            vkFreeMemory(device.getDevice(), imageMemory, null);
        }
        destroyed = true;
    }
}

File: src/cxx/VulkanLib/Device/Image/Image.cpp
//
// Created by kgaft on 3/11/24.
//
#include "Image.hpp"

Image::Image(std::shared_ptr<LogicalDevice> device, vk::Image base) : device(std::move(device)), base(base) {
    castCreated = true;
}

Image::Image(std::shared_ptr<LogicalDevice> &device, vk::ImageCreateInfo &createInfo) : imageInfo(createInfo) {
    initialize(device, createInfo);
}

Image::Image() {

}

std::shared_ptr<ImageView> Image::createImageView(vk::ImageViewCreateInfo &createInfo) {
    vk::ImageView view = Image::device->getDevice().createImageView(createInfo, VkLibAlloc::acquireAllocCb().get());
    imageViews.push_back(std::make_shared<ImageView>(imageInfo,
                                                     Image::device, view, createInfo));
    return imageViews[imageViews.size() - 1];
}

void Image::initialize(std::shared_ptr<LogicalDevice> device, vk::Image base) {
    Image::device = device;
    Image::base = base;
}

void Image::initialize(std::shared_ptr<LogicalDevice> &device, vk::ImageCreateInfo &createInfo) {
    vk::Result res;
    Image::device = device;
    Image::imageInfo = createInfo;
    base =  Image::device->getDevice().createImage(createInfo, VkLibAlloc::acquireAllocCb().get());

    vk::MemoryRequirements requirements;
    Image::device->getDevice().getImageMemoryRequirements(base, &requirements);

    vk::MemoryAllocateInfo allocInfo{};
    allocInfo.allocationSize = requirements.size;
    allocInfo.memoryTypeIndex =  Image::device->findMemoryType(requirements.memoryTypeBits,
                                                               vk::MemoryPropertyFlagBits::eDeviceLocal);

    res =  Image::device->getDevice().allocateMemory(&allocInfo, VkLibAlloc::acquireAllocCb().get(), &imageMemory);
    if(res!=vk::Result::eSuccess){
        throw std::runtime_error("Failed to allocate image memory");
    }
    Image::device->getDevice().bindImageMemory(base, imageMemory, 0);

}

vk::Image &Image::getBase() {
    return base;
}

void Image::copyFromBuffer(Buffer &buffer, uint32_t layerCount, LogicalQueue &queue) {
    vk::CommandBuffer cmd = queue.beginSingleTimeCommands();
    transitionImageLayout(cmd, imageInfo.initialLayout, vk::ImageLayout::eTransferDstOptimal, vk::ImageAspectFlagBits::eColor);
    vk::BufferImageCopy region{};
    region.bufferOffset = 0;
    region.bufferRowLength = 0;
    region.bufferImageHeight = 0;
    region.imageSubresource.aspectMask = vk::ImageAspectFlagBits::eColor;
    region.imageSubresource.mipLevel = 0;
    region.imageSubresource.baseArrayLayer = 0;
    region.imageSubresource.layerCount = layerCount;
    region.imageOffset.x = 0;
    region.imageOffset.y = 0;
    region.imageOffset.z = 0;
    region.imageExtent = imageInfo.extent;
    cmd.copyBufferToImage(buffer.getBuffer(), base, vk::ImageLayout::eTransferDstOptimal, 1, &region);
    transitionImageLayout(cmd,vk::ImageLayout::eTransferDstOptimal,  vk::ImageLayout::eGeneral, vk::ImageAspectFlagBits::eColor);
    imageInfo.initialLayout = vk::ImageLayout::eGeneral;
    queue.endSingleTimeCommands(cmd);
}

void Image::transitionImageLayout(std::shared_ptr<LogicalDevice> device, vk::ImageLayout oldLayout,
                                  vk::ImageLayout newLayout, vk::ImageAspectFlags aspectFlags) {
    vk::CommandBuffer commandBuffer = device->getQueueByType(vk::QueueFlagBits::eGraphics)->beginSingleTimeCommands();

    transitionImageLayout(commandBuffer, oldLayout, newLayout, aspectFlags);

    device->getQueueByType(vk::QueueFlagBits::eGraphics)->endSingleTimeCommands(commandBuffer);
}

void Image::transitionImageLayout(vk::CommandBuffer commandBuffer, vk::ImageLayout oldLayout, vk::ImageLayout newLayout,
                                  vk::ImageAspectFlags aspectFlags) {
    VkImageMemoryBarrier barrier{};
    barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
    barrier.oldLayout = static_cast<VkImageLayout>(oldLayout);
    barrier.newLayout = static_cast<VkImageLayout>(newLayout);
    barrier.srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
    barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
    barrier.image = base;
    barrier.subresourceRange.aspectMask = (unsigned int)aspectFlags;
    barrier.subresourceRange.baseMipLevel = 0;
    barrier.subresourceRange.levelCount = 1;
    barrier.subresourceRange.baseArrayLayer = 0;
    barrier.subresourceRange.layerCount = 1;

    VkPipelineStageFlags sourceStage;
    VkPipelineStageFlags destinationStage;

    if (static_cast<VkImageLayout>(oldLayout) == VK_IMAGE_LAYOUT_UNDEFINED && static_cast<VkImageLayout>(newLayout) == VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL)
    {
        barrier.srcAccessMask = 0;
        barrier.dstAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;

        sourceStage = VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT;
        destinationStage = VK_PIPELINE_STAGE_TRANSFER_BIT;
    }
    else if (static_cast<VkImageLayout>(oldLayout) == VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL &&
             static_cast<VkImageLayout>(newLayout) == VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL)
    {
        barrier.srcAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
        barrier.dstAccessMask = VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT;

        sourceStage = VK_PIPELINE_STAGE_TRANSFER_BIT;
        destinationStage = VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
    }
    else
    {
        barrier.srcAccessMask = 0;
        barrier.dstAccessMask = VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT;

        sourceStage = VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT;
        destinationStage = VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
    }

    vkCmdPipelineBarrier(
            commandBuffer,
            sourceStage, destinationStage,
            0,
            0, nullptr,
            0, nullptr,
            1, &barrier);
}

std::vector<std::shared_ptr<ImageView>> &Image::getImageViews() { return imageViews; }

void Image::resize(uint32_t width, uint32_t height) {
    if (!castCreated) {
        destroy();
        destroyed = false;
        imageInfo.extent = vk::Extent3D{width, height, 1};
        if(imageInfo.initialLayout!=vk::ImageLayout::eUndefined){
            vk::ImageLayout tmpLayout = imageInfo.initialLayout;
            imageInfo.initialLayout = vk::ImageLayout::eUndefined;
            initialize(device, imageInfo);
            transitionImageLayout(device, vk::ImageLayout::eUndefined, tmpLayout, vk::ImageAspectFlagBits::eColor);
            imageInfo.initialLayout = tmpLayout;
        } else {
            initialize(device, imageInfo);
        }
        for (auto &item: imageViews) {
            item->createInfo.image = base;
            item->base = device->getDevice().createImageView(item->createInfo, VkLibAlloc::acquireAllocCb().get());
            item->parentInfo = imageInfo;
        }
    }

}

vk::ImageCreateInfo &Image::getImageInfo() {
    if (castCreated) {
        throw std::runtime_error("Error: you cannot get image info, from image created via cast");
    }
    return imageInfo;
}

void Image::destroy() {
    if (!castCreated) {
        for (auto &item: imageViews) {
            item->destroy();
        }
        device->getDevice().destroyImage(base, VkLibAlloc::acquireAllocCb().get());
        device->getDevice().freeMemory(imageMemory, VkLibAlloc::acquireAllocCb().get());
    }
    destroyed = true;
}
